### DISCUSSION TITLE
AlpineJS x Sprig

### DISCUSSION THREAD

**Discussion by ishetnogferre** [alpinejs/alpine#3462] [7. Bugs (archived)]
For most of our Craft CMS, we're using both AlpineJS as Sprig to make things reactive. In the past we've had no problems when following this setup in the Sprig documentation: https://putyourlightson.com/plugins/sprig#alpine-js

However for our future projects, we'd like to try and separate the javascript from our HTML (twig), mainly because our team is divided in Javascript and Craft CMS developers and so this can be maintained by right people.

In trying to reach that goal, I've setup a small test project to find a solution. Here I'm using just an [AlpineJS dropdown from the docs](https://alpinejs.dev/component/dropdown) and wrapped it in a sprig component. I've found out that it can work if we re-init the javascript code after sprig has done it's job like so:

```
import Alpine from "alpinejs"
import dropdown from "@/scripts/components/Dropdown.js";

function init() {
	window.Alpine = Alpine

	Alpine.data('dropdown', dropdown);

	Alpine.start()
}

htmx.on('htmx:afterSettle', function () {
	init();
});
```

And that works, except for one error:
![Screenshot 2023-03-13 at 17 00 48](https://user-images.githubusercontent.com/6649706/224758148-cfb1900d-20f2-42be-b665-76e7e7f12968.png)

The piece of code where the error is thrown:

![Screenshot 2023-03-13 at 17 01 04](https://user-images.githubusercontent.com/6649706/224758155-47d65fa7-8857-4c1a-94b7-9d081f2a5d67.png)

Is there something we need to change in our code to make this work or something fixable in the Alpine code, or just not possible to use both frameworks together?

Thanks!

**Comment by ekwoka** [comment]
The issue here is that you'll end up starting multiple instances of alpine every time htmx settled, but the old ones will still be running.

I'd imagine what you're trying to figure out the way to do is the injecting of the component data before the component is registered in the dom when doing HTML updates, right?

What you'd need to do is have them just be global functions, that you add to the dom/evaluate BEFORE the dom containing the related `x-data` is added to the document.

Many ways to do this, one of which could be having the injected components having x-ignore and maybe their `x-data` is actually like `data-x-data`, and you can loop through them at the end of the injected script and swap the attributes and remove x-ignore (maybe in the opposite order? or manually calling initTree?) so the component can initialize later.

Alpine is just not made to have data components injected after Alpine has started.

**Reply by ishetnogferre** [reply]
I wanna destroy and (re-)init Alpine altogether after the htmx has settled, but couldn't find any particular docs for it, so tried to re-run the above init() function. Will take a look into what the initTree exactly is and if that can help, thanks!

**Reply by joshhanley** [reply]
@ishetnogferre have a look at how Livewire and Alpine work together, maybe you can get some inspiration from that as to how you can get it running. https://github.com/livewire/livewire/blob/master/js/component/SupportAlpine.js
